[metadata] Generate type-3 GUIDs for interfaces.

rbernon opened the pull request:
I'm opening this although it's still WIP, for discussion, as I'm not sure what else I should be adding. Tests probably, of course, but also I was wondering if it should be plugged to RuntimeType.GUID / Marshal.GenerateGuidForType. I'm not entirely sure if that's desired as I've seen clues in the reference source that the GUID generation has issues.

I've left class GUID generation aside, as it doesn't seem so useful for COM interop, and the algorithm uses some Assembly properties to seed the GUID, and retrieving that is currently beyond my understanding of the code.

Also I'm a bit lost regarding which RuntimeType.GUID it should be added to: there's at least two places that looked appropriate : in `mono/netcore/System.Private.CoreLib/src/RuntimeType.Mono.cs` and in `mono/mcs/class/corlib/ReferenceSources/RuntimeType.cs`.

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/mono/mono?pullRequest=20499) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/mono/mono?pullRequest=20499) before we can accept your contribution.<br/><br/>:x: rbernon [sign now](https://cla.dotnetfoundation.org/mono/mono?pullRequest=20499)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/mono/mono?pullRequest=20499) it.</sub>

lambdageek commented:
@rbernon thanks for the PR.

Couple of observations and questions:
1. Is there a bug that this is fixing?  could you summarize (either her or in a new GH Issue) the expected and actual behavior.
2. With this PR, if I understand it, for a class that has a `[Guid]` attribute, the code continues to use the provided guid.  For classes that don't have one, it creates a string with a representation of the class and all its methods and then hashes that and creates a GUID from the hash. 
   - Would it be possible to do this mostly in managed code using reflection?
   - If not, is it possible to reuse some of the other C code for converting types or methods to strings (for example `mono_method_full_name` or `mono_type_get_name_full`).
3. Have you looked at how CoreCLR does this?  Particularly for https://github.com/dotnet/runtime we're trying to converge as much as possible for stuff like this - ideally sharing managed C# code to implement similar functionality.

To answer your questions:
- `RuntimeType.GUID` and `Marshal.GenerateGuidForType` - we should identify gaps between what .NET Framework can do and what Mono can do.  If there are cases where Mono gives a different answer, we should try to patch up the behavior.

- About the duplicated .cs files:
  - `mono/netcore/System.Private.CoreLib/src/RuntimeType.Mono.cs` this is the "netcore" version of Mono's base class library - this is pretty much obsolete.  New changes shoudl go into https://github.com/dotnet/runtime into the `src/mono/netcore/System.Private.CoreLib` directory.
 - `mono/mcs/class/corlib/ReferenceSources/RuntimeType.cs` - this is the .NET Framework-compatible version of Mono's base class library (is our mscorlib.dll).  This is generally what you want to change in https://github.com/mono/mono.


(There is a bot that runs and mirrors pull requests C source changes from https://github.com/mono/mono `src/mono` to https://github.com/dotnet/runtime `src/mono/mono`.)

rbernon commented:
> Couple of observations and questions:
> 
>     1. Is there a bug that this is fixing?  could you summarize (either her or in a new GH Issue) the expected and actual behavior.

Yes sorry I should have described that a bit more in detail. The issue happens with L.A.Noire game (running with Wine), that loads a .NET assembly on startup for its error reporting mechanism.

The interop is done through COM Callable Wrapper, and the game instantiates the class using CoCreateInstance (that resolves the class and the assembly from the registry), and then calls `QueryInterface` on the returned wrapper with the interface generated IID that it has embedded (which probably comes from the typelib that `tlbexp.exe` can generate from the assembly).

Mono currently doesn't implement the generated GUIDs, so the `QueryInterface` call fails to find the corresponding interface, and the game crashes.

>     2. With this PR, if I understand it, for a class that has a `[Guid]` attribute, the code continues to use the provided guid.  For classes that don't have one, it creates a string with a representation of the class and all its methods and then hashes that and creates a GUID from the hash.
>        
>        * Would it be possible to do this mostly in managed code using reflection?

Probably, although the interface lookup was done in the cominterop code, and it's already doing the GUID attribute lookup in native code. It could all be done in managed code, if only I knew how to call it from the native cominterop. 

>        * If not, is it possible to reuse some of the other C code for converting types or methods to strings (for example `mono_method_full_name` or `mono_type_get_name_full`).
> 

The serialization has to follow a specific format for the GUID to be generated in the same way as `tlbexp.exe` does, otherwise they won't match. If those methods return the right format then possibly we can re-use them yes (although the method names apparently aren't taken into consideration).

>     3. Have you looked at how CoreCLR does this?  Particularly for https://github.com/dotnet/runtime we're trying to converge as much as possible for stuff like this - ideally sharing managed C# code to implement similar functionality.
> 

Yes I actually re-implemented the algorithm from there (https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/methodtable.cpp#L7300 and related sources), as it has to match. I verified that the re-implementation works with what the game needs, but it's maybe also a simple use-case, I was planning on adding more tests. But then the question was wether that had to be plugged to RuntimeType.GUID or not. That could change the way those tests would be written.

marek-safar commented:
approve

marek-safar commented:
> But then the question was wether that had to be plugged to RuntimeType.GUID or not. 

It should, RuntimeType.GUID should return GuidAttribute value or the pseudo-GUID for the type.

rbernon commented:
So, should I be doing that in managed code instead? In `mono/mcs/class/corlib/ReferenceSources/RuntimeType.cs` only?

TBH I'm much more comfortable with the C code, it's been quite straightforward and I can easily see how to call it from the managed code.

marek-safar commented:
> So, should I be doing that in managed code instead?

Calling into native code is fine for this

lambdageek commented on mono/metadata/metadata.c:
```diff
@@ -7824,3 +7826,262 @@ mono_type_set_amods (MonoType *t, MonoAggregateModContainer *amods)
+		return;
+	}
+	if (klass->nested_in) {
+		mono_signature_append_class_name (res, klass->nested_in);
```
```suggestion
		mono_signature_append_class_name (res, m_class_get_nested_in (klass));
```

We try not to access the fields of `MonoClass` directly (see `mono/metadata/class-getters.h`).

lambdageek commented on mono/metadata/metadata.c:
```diff
@@ -7824,3 +7826,262 @@ mono_type_set_amods (MonoType *t, MonoAggregateModContainer *amods)
+	mono_md5_update (&ctx, (guchar *)unicode_name, items_written * sizeof(gunichar2));
+
+	g_free (unicode_name);
+	g_string_free (name, FALSE);
```
```suggestion
	g_string_free (name, TRUE);
```

Usually if you pass `FALSE`, you then need to take the return value (which is the value of `name->str`) and use it for something or free it yourself.

lambdageek commented on mono/metadata/metadata.c:
```diff
@@ -7824,3 +7826,262 @@ mono_type_set_amods (MonoType *t, MonoAggregateModContainer *amods)
+		GString *res = g_string_new ("");
+		mono_guid_signature_append_method (res, sig);
+		mono_md5_update (&ctx, (guchar *)res->str, res->len);
+		g_string_free (res, FALSE);
```
```suggestion
		g_string_free (res, TRUE);
```

lambdageek reviewed (approved):
LGTM.  There's a couple of memory leaks from `g_string_free` that should be fixed up.

lambdageek commented on mono/metadata/metadata.c:
```diff
@@ -7824,3 +7826,262 @@ mono_type_set_amods (MonoType *t, MonoAggregateModContainer *amods)
 	t_full->mods.amods = amods;
 }
+
+static void
```
Probably everything here should be under `#ifndef DISABLE_COM` for the configurations where we don't compile in the COM support.  (Pretty much all the AOT platforms, for example)

rbernon replied:
I'm going to make RuntimeType.GUID use it too, so I suppose that's not going to be only for COM anymore.

marek-safar replied:
That's not what we want, for COM disabled configurations it needs to return empty GUID

lambdageek commented on mono/metadata/metadata.h:
```diff
@@ -512,6 +512,8 @@ MONO_API uint32_t      mono_metadata_decode_table_row_col (MonoImage *image, int
 					   int            idx,
 					   unsigned int          col);
 
+MONO_API void  mono_metadata_get_class_guid (MonoClass* klass, uint8_t* guid, MonoError *error);
```
The `metadata.h` header is part of Mono's public API.  We shouldn't add stuff to it without good reason.

1. Move this declaration to `metadata-internals.h` (or `cominterop.h`)
2. Remove the `MONO_API` attribute.

rbernon commented:
I've updated the commit, addressing the few comments that were made and implementing the whole thing more completely. I'd say it's not much of a WIP now, and more of a pull request.

I added a few test case to check that the generated GUID can now be used to query interfaces and that they match what .NET Framework does on Windows (and found some bug in the serialization with `in` parameters, that I had to reproduce. I think https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/utilcode/prettyprintsig.cpp#L691 doesn't even have it right, but I'm maybe misreading the code there.

Also fixed a mistake with interfaces living in the global namespace (empty namespace name wasn't checked correctly).

The generated GUID is still only implemented for interfaces. I don't think it's much useful for classes, and according to the reference code it would require introspection of the class assembly.

lambdageek commented on mono/metadata/icall.c:
```diff
@@ -3275,6 +3275,28 @@ set_type_object_in_array (MonoDomain *domain, MonoType *type, MonoArrayHandle de
+	if (!mono_class_init_checked (klass, error))
+		return;
+
+	MONO_ENTER_NO_SAFEPOINTS;
+	guint8 *data = (guint8*) mono_array_addr_with_size_internal (MONO_HANDLE_RAW (guid_handle), 1, 0);
+	mono_metadata_get_class_guid (klass, data, error);
+	MONO_EXIT_NO_SAFEPOINTS;
```
```suggestion
	guint8 *data = (guint8*) mono_array_addr_with_size_internal (MONO_HANDLE_RAW (guid_handle), 1, 0);
	mono_metadata_get_class_guid (klass, data, error);
```

You don't need the "no safepoints" boundary here.  We recently changed how our local managed object handles work - they prevent the GC from moving the object around.  So you can just assume that `data` will be in the same place until the icall returns.


(and it probably isn't correct, because I think pretty printing the interface type may trigger assembly loading, which may trigger running managed code which could sfepoint).

We recent

rbernon replied:
I updated the commit, removing `MONO_ENTER_NO_SAFEPOINTS`/`MONO_EXIT_NO_SAFEPOINTS`

rbernon commented:
Hi, gentle ping to be sure the update didn't get unnoticed. Anything more I should perhaps?

marek-safar commented:
@lambdageek could you help with landing this PR

rbernon commented:
Hello, anything I can do to help here?

lambdageek commented:
Hey, sorry about that.  I'm going to push a commit to fix the build problems:
1. a few places where the MonoClass members are accessed directly without using a getter.
2. An OSX build failure where we use the [`CommonCrypto/CommonDigest.h`](https://github.com/Apple-FOSS-Mirror/CommonCrypto/blob/master/CommonCrypto/CommonDigest.h) instead of our own md5 implementation - just need another way to get at the digest length

I expect to merge the PR today

lambdageek commented:
@rbernon I don't have permission to push to your branch, but you should pick up these two commits https://github.com/rbernon/mono/compare/cominterop-generated-guid...lambdageek:cominterop-generated-guid

rbernon commented:
Thanks! I've rebased and force pushed the branch with your additional commits.

lambdageek commented:
Update: it's taking longer than I hoped to get the new test to build on Win32.  (my attempts over in https://github.com/dotnet/runtime/pull/45900).  I should have something in a few hours.

lambdageek commented:
Please take 2 more commits from https://github.com/rbernon/mono/compare/cominterop-generated-guid...lambdageek:cominterop-generated-guid

lambdageek commented:
@monojenkins build failed

rbernon commented:
Thanks!
